Add GCC 12 CI configuration

msimberg opened the pull request:
Fixes #839.

msimberg commented:
cscs-ci run

msimberg commented:
cscs-ci run

msimberg commented:
cscs-ci run

msimberg commented:
https://github.com/eth-cscs/DLA-Future/pull/853/commits/b173a1f83e4b630011a920aa4223e76ddfc2f8db works around a bug in GCC that seems to appear in GCC 12... and GCC 13 prerelease available in Ubuntu 23.04. See here for a minimal example: https://godbolt.org/z/KP4W689Pb. Clang seems to have no trouble with the example with any version. `gcc-13` in Ubuntu 23.04 reports as its version
```
gcc-13 (Ubuntu 13-20230320-1ubuntu1) 13.0.1 20230320 (experimental) [master r13-6759-g5194ad1958c]
```
so I'm guessing the fix was maybe not there yet in that version (GCC 13 on compiler explorer has no problems).

msimberg commented:
cscs-ci run

msimberg commented:
cscs-ci run

RMeli commented on ci/.gitlab-ci.yml:
```diff
@@ -134,6 +156,28 @@ cpu release build clang13 cxx20:
+    DEPLOY_DOCKER_FILE: ci/docker/deploy.Dockerfile
+    BASE_IMAGE: docker.io/ubuntu:23.04
+    DEPLOY_BASE_IMAGE: docker.io/ubuntu:23.04
+    EXTRA_APTGET: clang-16
```
```suggestion
    EXTRA_APTGET: "clang-16 libomp-16-dev"
```
We can do this here or in #668

msimberg replied:
Good point. I think we can have #668 in first. I'll update this here already now.

msimberg replied:
https://github.com/eth-cscs/DLA-Future/pull/853/commits/3ee241fc7295b4b2143e9996b47ff118cfa25efc

msimberg commented:
Waiting for #668 to be merged before this.

msimberg commented:
cscs-ci run

msimberg commented:
cscs-ci run

msimberg commented:
It seems like the problem of `blaspp` not configuring with clang 16 (https://gitlab.com/cscs-ci/ci-testing/webhook-ci/mirrors/4700071344751697/7514005670787789/-/jobs/4265672150#L1368) was due to the linker not finding `libomp` (even though `libomp-16-dev` was installed). From exploring a bit it seems like for clang 13 and `libomp-13-dev` the linker explicitly adds a path to an llvm directory which contains `libomp.so`. With clang 14, 15, and 16 that path is not set by the linker so it doesn't find `libomp.so`. Instead of trying to patch this by setting linker paths manually I tried the LLVM repository and there everything seems to be correctly configured. So I switched the base image back to 22.04 for clang and I'm now using the LLVM repo instead for the clang 16 configuration.

Because I changed the clang 16 configuration back to 22.04 I also changed the GCC 13 configuration to GCC 12 on top of 22.04. Now all configurations are again on the LTS release. Additionally, the GCC 13 in 23.04 was anyway a pre-release so it's maybe better to stick with 12 just because of that as well.

Once I've figured out what I messed up in the configuration file syntax this should be ready to go again...

msimberg commented:
cscs-ci run

msimberg commented:
I need more investigation on the clang behaviour so I've removed the clang 16 configuration from this PR. The state before removing the clang 16 configuration is still here: https://github.com/eth-cscs/DLA-Future/compare/master...msimberg:DLA-Future:clang-16-ci-config.

My previous investigation was sloppy: with the packages from Ubuntu's repositories I couldn't even compile with `clang++ -fopenmp file.cpp`, much less `clang++ -lomp file.cpp`. With clang from LLVM's repositories I tested that `clang++ -fopenmp file.cpp` worked. However, `clang++ -lomp file.cpp` did not work which is why CI was still failing. Because of https://github.com/spack/spack/blob/bcb7af6eb371b8958824e2f0551c4baa88d754a3/lib/spack/spack/build_systems/intel.py#L704 mkl is adding `-lomp` to `blas_libs`, not `-fopenmp` (or `-fopenmp=libomp`). I don't know if that's correct or if that should be fixed in spack. I'm going to see how a clang-built clang 16 behaves separately from this PR.

msimberg commented:
cscs-ci run
